misk/src/main/kotlin/misk/scope/executor/ActionScopedExecutorService.kt

```diff
@@ -0,0 +1,24 @@
+package misk.scope.executor
+
+import com.google.common.util.concurrent.ListeningExecutorService
+import com.google.common.util.concurrent.MoreExecutors
+import misk.concurrent.WrappingListeningExecutorService
+import misk.scope.ActionScope
+import java.util.concurrent.Callable
+import java.util.concurrent.ExecutorService
+
+/**
+ * Wraps a [ListeningExecutorService] to propagate the current action scope to any tasks
+ * submitted by the current thread
+ */
+class ActionScopedExecutorService(
```

Comment:
I love how the abstractions layer for things like this.

That said, I’m still anxious about the application-layer consequences of scoped objects being shared between threads

Reply:
... would we be better off encouraging more of an Actors approach – if your action wants to do work in another thread you can‘t share state. 

I think the most valuable state that we transition with this is trace IDs. But I think trace IDs might be the exception and not the rule. Instead of trampolining all possible action-scoped stuff into other threads; maybe we should special case trace IDs.

Trace IDs are also the only things that effortlessly trampoline between processes. They’re not like application-layer things like HTTP request headers.

Reply:
I think many things propagate - calling principal for example should be inherited across threads spawned by the action, and should propagate across process boundaries as a principal chain (principal acting on-behalf of another principal). Not sure where the tipping point is though.